controllers/datadoghq/tls.go

```diff
@@ -0,0 +1,232 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package datadoghq
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/DataDog/watermarkpodautoscaler/apis/datadoghq/v1alpha1"
+)
+
+const (
+	certificateCacheTimeout    = 10 * time.Minute
+	certificateCacheLRUTimeout = 10 * time.Minute
+)
+
+type tlsTransport struct {
+	wrappedTransport *http.Transport
+}
+
+// tlsCertificateCache is a LRU cache to store in memory certificates
+// used for TLS connections
+type tlsCertificateCache struct {
+	mu    sync.RWMutex
+	cache map[string]*tlsCacheEntry
+}
+
+// tlsCacheEntry is an entry of the certificate cache
+type tlsCacheEntry struct {
+	certificate *tls.Certificate
+	lastUpdate  time.Time
+	lastAccess  time.Time
+}
+
+func newTLSCertificateCache() *tlsCertificateCache {
+	cache := &tlsCertificateCache{
+		cache: make(map[string]*tlsCacheEntry),
+	}
+	go cache.run()
+	return cache
+}
+
+func (c *tlsCertificateCache) GetClientCertificateReloadingFunc(certFile, keyFile string) func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	return func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		now := time.Now()
+		c.mu.RLock()
+		entry, ok := c.cache[certFile]
+		c.mu.RUnlock()
+
+		// there's a small race condition here, but it does no harm except
+		// possibly reloading the certificate from disk more than once at a given time
+		if !ok || entry.isExpired(now) || entry.isCertificateExpired(now) {
+			certificate, err := retryLoadingX09Keypair(5, 50*time.Millisecond, certFile, keyFile)
```

Comment:
Could we kill the WPA here if we continusouly reload a bad certificate a lot ?

Reply:
Yes, good idea.